libs/vkd3d-shader/spirv.c

```diff
@@ -7220,6 +7395,81 @@ static const struct vkd3d_symbol *vkd3d_dxbc_compiler_find_resource(struct vkd3d
+    uint32_t uint_type_id, ptr_type_id, ptr_id, indices[2];
+
+    uint_type_id = vkd3d_spirv_get_type_id(builder, VKD3D_TYPE_UINT, 1);
+    ptr_type_id = vkd3d_spirv_get_op_type_pointer(builder, SpvStorageClassPushConstant, uint_type_id);
+
+    indices[0] = vkd3d_dxbc_compiler_get_constant_uint(compiler, compiler->descriptor_table_member);
+    indices[1] = vkd3d_dxbc_compiler_get_constant_uint(compiler, descriptor_table);
+
+    ptr_id = vkd3d_spirv_build_op_access_chain(builder, ptr_type_id,
+            compiler->descriptor_table_var_id, indices, ARRAY_SIZE(indices));
+
+    return vkd3d_spirv_build_op_load(builder, uint_type_id, ptr_id, SpvMemoryAccessMaskNone);
+}
+
+static void vkd3d_dxbc_compiler_decorate_nonuniform(struct vkd3d_dxbc_compiler *compiler,
+        uint32_t expression_id)
+{
+    struct vkd3d_spirv_builder *builder = &compiler->spirv_builder;
+
+    vkd3d_spirv_enable_capability(builder, SpvCapabilityShaderNonUniformEXT);
+    vkd3d_spirv_build_op_decorate(builder, expression_id, SpvDecorationNonUniformEXT, NULL, 0);
+}
+
+static uint32_t vkd3d_dxbc_compiler_get_resource_pointer(struct vkd3d_dxbc_compiler *compiler,
+        const struct vkd3d_shader_register *reg)
+{
+    struct vkd3d_spirv_builder *builder = &compiler->spirv_builder;
+    uint32_t ptr_id, ptr_type_id, index_id;
+    const struct vkd3d_symbol *symbol;
+
+    symbol = vkd3d_dxbc_compiler_find_resource(compiler, reg);
+    ptr_id = symbol->id;
+
+    if (symbol->info.resource.resource_binding->flags & VKD3D_SHADER_BINDING_FLAG_BINDLESS)
+    {
+        unsigned int descriptor_table = symbol->info.resource.resource_binding->descriptor_table;
+        unsigned int descriptor_index = symbol->info.resource.resource_binding->descriptor_offset
+                    - symbol->info.resource.resource_binding->register_index;
+
+        if (shader_is_sm_5_1(compiler))
+        {
+            struct vkd3d_shader_register_index index = reg->idx[1];
+            index.offset += descriptor_index;
+            index_id = vkd3d_dxbc_compiler_emit_register_addressing(compiler, &index);
+        }
+        else
+        {
+            index_id = vkd3d_dxbc_compiler_get_constant_uint(compiler,
+                    descriptor_index + reg->idx[0].offset);
+        }
+
+        index_id = vkd3d_spirv_build_op_iadd(builder,
+                vkd3d_spirv_get_type_id(builder, VKD3D_TYPE_UINT, 1),
+                vkd3d_dxbc_compiler_load_descriptor_table_offset(compiler, descriptor_table),
+                index_id);
+
+        ptr_type_id = vkd3d_spirv_get_op_type_pointer(builder,
+                symbol->info.resource.storage_class,
+                symbol->info.resource.type_id);
+
+        ptr_id = vkd3d_spirv_build_op_access_chain(builder,
+                ptr_type_id, ptr_id, &index_id, 1);
+    }
+
+    /* AMD drivers rely on the index being marked as nonuniform */
+    if (reg->modifier == VKD3DSPRM_NONUNIFORM)
```

Comment:
This is somewhat sketchy since index_id can be uninitialized. Shouldn't this only happen inside the branch?

Reply:
Fixed.